presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -15,259 +15,196 @@
+ * until pages are fetched by caller to reduce the buffer under its memory cap.
+ * 4. Caller must call stop() to release resource.
  */
 public class NativeExecutionTask
 {
     private static final Logger log = Logger.get(NativeExecutionTask.class);
-    private static final int GET_TASK_INFO_INTERVALS = 100;
     private static final String LOCAL_URI = "http://127.0.0.1/v1/task/";
 
     private final Session session;
     private final TaskId taskId;
     private final PlanFragment planFragment;
     private final OutputBuffers outputBuffers;
-    private final HttpClient httpClient;
+    private final PrestoSparkHttpWorkerClient workerClient;
     private final TableWriteInfo tableWriteInfo;
-    private final RemoteTaskStats stats;
     private final List<TaskSource> sources;
     private final Executor executor;
     private final ScheduledExecutorService updateScheduledExecutor;
-    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
-    private final JsonCodec<PlanFragment> planFragmentCodec;
-    private final JsonCodec<TaskInfo> taskInfoCodec;
-    private final AtomicReference<TaskInfo> taskInfo = new AtomicReference<>();
-
-    @GuardedBy("this")
-    private ScheduledFuture<?> scheduledFuture;
+    private final HttpNativeExecutionTaskInfoFetcher taskInfoFetcher;
+    private final HttpNativeExecutionTaskResultFetcher taskResultFetcher;
 
     public NativeExecutionTask(
             Session session,
+            URI location,
             TaskId taskId,
             PlanFragment planFragment,
             List<TaskSource> sources,
             HttpClient httpClient,
             TableWriteInfo tableWriteInfo,
-            RemoteTaskStats stats,
             Executor executor,
-            ScheduledExecutorService updateScheduledExecutor,
-            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
-            JsonCodec<PlanFragment> planFragmentCodec,
-            JsonCodec<TaskInfo> taskInfoCodec)
+            ScheduledExecutorService updateScheduledExecutor)
     {
         requireNonNull(session, "session is null");
         requireNonNull(taskId, "taskId is null");
         requireNonNull(planFragment, "planFragment is null");
         requireNonNull(sources, "sources is null");
         requireNonNull(httpClient, "httpClient is null");
         requireNonNull(tableWriteInfo, "tableWriteInfo is null");
-        requireNonNull(stats, "stats is null");
         requireNonNull(executor, "executor is null");
         requireNonNull(updateScheduledExecutor, "updateScheduledExecutor is null");
-        requireNonNull(taskUpdateRequestCodec, "taskUpdateRequestCodec is null");
-        requireNonNull(planFragmentCodec, "planFragmentCodec is null");
-        requireNonNull(taskInfoCodec, "taskInfoCodec is null");
 
         try (SetThreadName ignored = new SetThreadName("NativeExecutionTask-%s", taskId)) {
```

Comment:
I don't know if we need this or not because the thread is just doing a lot of assignments